### PR TITLE
Abort cmov optimization, when one of the out blocks is try block

### LINKED ISSUE #17733: Assertion at mini.c:2644, condition `tblock->native_offset' not met
## Steps to Reproduce

1. Compile the code below as `csc KillMono.cs`.
2. Run it as `mono KillMono.exe` and it crashes.
3. Interestingly, a binary generated by `mcs KillMono.cs` does not kill the mono runtime.

KillMono.cs:
```csharp
using System;

class Debug
{
	public static string test(string s)
	{   
		if (s.Length == 0) return "Empty";

		try {return s;}
		catch (Exception) {}

		return "Should not happen";
	}

	public static void Main()
	{
		test("abc");
	}
}
```
(Of course try-catch in this code is non-sense but the runtime should not crash.)

### Current Behavior

Mono crashes with the following error.

```
* Assertion at mini.c:2644, condition `tblock->native_offset' not met

=================================================================
	Native Crash Reporting
=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================
```

### Expected Behavior

Print nothing and exit cleanly.

## On which platforms did you notice this

Linux

**Version Used**:

- Mono JIT compiler version 6.4.0.198
- Microsoft (R) Visual C# Compiler version 3.3.1-beta4-19462-11 (66a912c9)
- Mono C# compiler version 6.4.0.198

It also happens with

- Mono JIT compiler version 5.18.1.0

## Stacktrace
```
  Id   Target Id         Frame 
  3    Thread 0x7f10773ff700 (LWP 784068) "SGen worker" 0x00007f107e4b4965 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  2    Thread 0x7f1074cc9700 (LWP 784069) "Finalizer" 0x00007f107e4b6adb in do_futex_wait.constprop.1 () from /lib64/libpthread.so.0
* 1    Thread 0x7f107f1d6780 (LWP 784067) "mono" 0x00007f107e4b8179 in waitpid () from /lib64/libpthread.so.0

Thread 3 (Thread 0x7f10773ff700 (LWP 784068)):
#0  0x00007f107e4b4965 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00000000006f23b3 in mono_os_cond_wait (mutex=0xa86800 <lock>, cond=0xa867c0 <work_cond>) at ../../mono/utils/mono-os-mutex.h:177
#2  get_work (job=<synthetic pointer>, do_idle=<synthetic pointer>, work_context=<synthetic pointer>, worker_index=0) at sgen-thread-pool.c:165
#3  thread_func (data=<optimized out>) at sgen-thread-pool.c:196
#4  0x00007f107e4b0dd5 in start_thread () from /lib64/libpthread.so.0
#5  0x00007f107dfc402d in clone () from /lib64/libc.so.6

Thread 2 (Thread 0x7f1074cc9700 (LWP 784069)):
#0  0x00007f107e4b6adb in do_futex_wait.constprop.1 () from /lib64/libpthread.so.0
#1  0x00007f107e4b6b6f in __new_sem_wait_slow.constprop.0 () from /lib64/libpthread.so.0
#2  0x00007f107e4b6c0b in sem_wait@@GLIBC_2.2.5 () from /lib64/libpthread.so.0
#3  0x00000000006970f4 in mono_os_sem_wait (flags=MONO_SEM_FLAGS_ALERTABLE, sem=0xa779e0 <finalizer_sem>) at ../../mono/utils/mono-os-semaphore.h:203
#4  mono_coop_sem_wait (flags=MONO_SEM_FLAGS_ALERTABLE, sem=0xa779e0 <finalizer_sem>) at ../../mono/utils/mono-coop-semaphore.h:41
#5  finalizer_thread (unused=unused@entry=0x0) at gc.c:967
#6  0x000000000064cc1b in start_wrapper_internal (stack_ptr=<optimized out>, start_info=0x0) at threads.c:1221
#7  start_wrapper (data=0x226f6b0) at threads.c:1294
#8  0x00007f107e4b0dd5 in start_thread () from /lib64/libpthread.so.0
#9  0x00007f107dfc402d in clone () from /lib64/libc.so.6

Thread 1 (Thread 0x7f107f1d6780 (LWP 784067)):
#0  0x00007f107e4b8179 in waitpid () from /lib64/libpthread.so.0
#1  0x00000000004b9d7f in dump_native_stacktrace (mctx=mctx@entry=0x7ffe68f8e540, signal=0x732b08 "SIGABRT") at mini-posix.c:1110
#2  0x00000000004b9f21 in mono_dump_native_crash_info (signal=signal@entry=0x732b08 "SIGABRT", mctx=mctx@entry=0x7ffe68f8e540, info=info@entry=0x7ffe68f8e830) at mini-posix.c:1158
#3  0x00000000004688c9 in mono_handle_native_crash (signal=signal@entry=0x732b08 "SIGABRT", mctx=mctx@entry=0x7ffe68f8e540, info=info@entry=0x7ffe68f8e830) at mini-exceptions.c:3328
#4  0x00000000004b92cf in sigabrt_signal_handler (_dummy=6, _info=0x7ffe68f8e830, context=0x7ffe68f8e700) at mini-posix.c:234
#5  <signal handler called>
#6  0x00007f107defc2c7 in raise () from /lib64/libc.so.6
#7  0x00007f107defd9b8 in abort () from /lib64/libc.so.6
#8  0x00000000007156d4 in monoeg_assert_abort () at goutput.c:57
#9  0x00000000006fc7d9 in mono_log_write_logfile (log_domain=<optimized out>, level=<optimized out>, hdr=<optimized out>, message=0x227d9a0 "* Assertion at mini.c:2644, condition `tblock->native_offset' not met\n") at mono-log-common.c:136
#10 0x00000000007155fe in monoeg_g_logstr (msg=<optimized out>, log_level=G_LOG_LEVEL_ERROR, log_domain=0x0) at goutput.c:134
#11 monoeg_g_logv_nofree (log_domain=log_domain@entry=0x0, log_level=log_level@entry=G_LOG_LEVEL_ERROR, format=format@entry=0x71e6d8 "* Assertion at %s:%d, condition `%s' not met\n", args=args@entry=0x7ffe68f8eed8) at goutput.c:149
#12 0x0000000000715ad0 in monoeg_g_logv_nofree (args=0x7ffe68f8eed8, format=0x71e6d8 "* Assertion at %s:%d, condition `%s' not met\n", log_level=G_LOG_LEVEL_ERROR, log_domain=0x0) at goutput.c:142
#13 monoeg_assertion_message (format=format@entry=0x71e6d8 "* Assertion at %s:%d, condition `%s' not met\n") at goutput.c:184
#14 0x00000000004bacc3 in create_jit_info (cfg=cfg@entry=0x227e620, method_to_compile=method_to_compile@entry=0x227f488) at mini.c:2644
#15 0x00000000004c0765 in mini_method_compile (method=method@entry=0x227f488, opts=opts@entry=374434303, domain=domain@entry=0x223d8b0, flags=flags@entry=JIT_FLAG_RUN_CCTORS, parts=parts@entry=0, aot_method_index=aot_method_index@entry=-1) at mini.c:3857
#16 0x00000000004c1252 in mono_jit_compile_method_inner (method=method@entry=0x227f488, target_domain=target_domain@entry=0x223d8b0, opt=opt@entry=374434303, error=error@entry=0x7ffe68f8f3e0) at mini.c:4035
#17 0x000000000042d6e4 in mono_jit_compile_method_with_opt (method=method@entry=0x227f488, opt=374434303, jit_only=jit_only@entry=0, error=error@entry=0x7ffe68f8f3e0) at mini-runtime.c:2445
#18 0x000000000042e7ac in mono_jit_compile_method (method=method@entry=0x227f488, error=error@entry=0x7ffe68f8f3e0) at mini-runtime.c:2494
#19 0x000000000046bff2 in common_call_trampoline (regs=regs@entry=0x7ffe68f8f4b8, code=code@entry=0x40548b18 "H\203\304\b\303\350ޖx\001\004\210\364'\002", m=m@entry=0x227f488, vt=vt@entry=0x0, vtable_slot=<optimized out>, vtable_slot@entry=0x0, error=error@entry=0x7ffe68f8f3e0) at mini-trampolines.c:637
#20 0x000000000046c983 in mono_magic_trampoline (regs=0x7ffe68f8f4b8, code=0x40548b18 "H\203\304\b\303\350ޖx\001\004\210\364'\002", arg=0x227f488, tramp=<optimized out>) at mini-trampolines.c:777
#21 0x0000000041cd2387 in ?? ()
#22 0x000000000225a315 in ?? ()
#23 0x00007f107df4e599 in __default_morecore () from /lib64/libc.so.6
#24 0x00007ffe68f8f4b8 in ?? ()
#25 0x00000000022662a0 in ?? ()
#26 0x0000000000a8c908 in sgen_nursery_max_size ()
#27 0x0000000040548af0 in ?? ()
#28 0x00007ffe68f8f758 in ?? ()
#29 0x00007ffe68f8f758 in ?? ()
#30 0x00007ffe68f8f680 in ?? ()
#31 0x00007ffe68f8f6f0 in ?? ()
#32 0x00007ffe68f8f8f0 in ?? ()
#33 0x00007f107f11c130 in ?? ()
#34 0x0000000000000000 in ?? ()
```

### PR DESCRIPTION
**Pull request by fanyang-mono**
Fixes #17733

### CODE REVIEW

File: mono/mini/iltests.il [hunk 1]
```diff
@@ -3491,4 +3491,74 @@ L3:
...
+	.method public hidebysig static int32 test_1_cmov_try_block() cil managed
+	{
+		// Code size       37 (0x25)
+		.maxstack  2
+		.locals init (string V_0, bool V_1, int32 V_2)
+		IL_0000:  nop
+		IL_0001:  ldstr  "abc"
+		IL_0006:  call   string Tests::cmov_try_block_helper(string)
+		IL_000b:  stloc.0
+		IL_000c:  ldloc.0
```
Comment (lewurm) [inline mono/mini/iltests.il]: isn't needed.

File: mono/mini/iltests.il [hunk 2]
```diff
@@ -3491,4 +3491,74 @@ L3:
...
+		IL_000d:  ldstr  "abc"
+		IL_0012:  call   bool [mscorlib]System.String::op_Equality(string,string)
+		IL_0017:  stloc.1
+		IL_0018:  ldloc.1
```
Comment (lewurm) [inline mono/mini/iltests.il]: isn't needed.

File: mono/mini/branch-opts.c [hunk 3]
```diff
@@ -321,6 +321,10 @@ mono_if_conversion (MonoCompile *cfg)
 		bb1 = bb->out_bb [0];
 		bb2 = bb->out_bb [1];
 
+		/* If either bb1 or bb2 is a try block, abort the optimization attempt. */
+		if (bb1->try_start || bb2->try_start)
+			continue;
```
Comment (lewurm) [inline mono/mini/branch-opts.c]: nit: Move it above the `restart:` label. The comment in line 307-317 refers to the `if` statement in line 328.

Comment (imhameed) [reply mono/mini/branch-opts.c]: Doesn't seem like a big deal? Moving this above the `restart:` label would also require that you supply some sort of bounds check for `out_bb` and initialization for `bb1` and `bb2` above `restart:`.

Alternatively, the comment could be moved instead.

Comment (fanyang-mono) [reply mono/mini/branch-opts.c]: Yes, I moved the comments instead.

**Comment by akoeplinger** [comment]
@monojenkins build

**Comment by akoeplinger** [comment]
@monojenkins build

**Pull request merged**